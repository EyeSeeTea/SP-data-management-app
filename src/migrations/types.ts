import { D2Api } from "../types/d2-api";

export interface Config {
    version: number;
    migration?: { version: number; error?: string };
}

export type Debug = (message: string) => void;

export interface MigrationWithVersion {
    version: number;
    migrate: MigrationFn;
    name: string;
}

export type Migration = Omit<MigrationWithVersion, "version">;

export type MigrationFn = (api: D2Api, debug: Debug) => Promise<void>;

export interface RunnerOptions {
    api: D2Api;
    debug?: Debug;
    dataStoreNamespace: string;
    migrations: MigrationWithVersion[];
}

export type MigrationTasks = MigrationWithVersion[];

export function migration(version: number, migration: Migration): MigrationWithVersion {
    return { version, ...migration };
}
