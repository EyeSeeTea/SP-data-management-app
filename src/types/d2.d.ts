declare module "d2/uid" {
    export function generateUid(): string;
}
