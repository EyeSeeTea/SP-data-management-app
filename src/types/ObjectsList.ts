export interface Pagination {
    page: number;
    pageSize: number;
    sorting: [string, "asc" | "desc"];
}
