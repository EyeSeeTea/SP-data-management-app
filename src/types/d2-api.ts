import { D2Api, D2ModelSchemas, MetadataPayloadBase } from "@eyeseetea/d2-api/2.36";
import MockAdapter from "axios-mock-adapter/types";

export * from "@eyeseetea/d2-api/2.36";

export function getMockApi(): { api: D2Api; mock: MockAdapter } {
    const api = new D2Api({ backend: "xhr" });
    const mock = api.getMockAdapter();
    return { api, mock };
}

export type D2Payload = Partial<MetadataPayloadBase<D2ModelSchemas>>;
