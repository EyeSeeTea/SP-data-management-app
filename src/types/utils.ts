export type Maybe<T> = T | undefined;

export type Either<Data, Error> = { type: "success"; data: Data } | { type: "error"; error: Error };

/* Like Partial<T>, but recursive on object values */
export type RecursivePartial<T> = {
    [P in keyof T]?: T[P] extends (infer U)[]
        ? RecursivePartial<U>[]
        : T[P] extends object
        ? RecursivePartial<T[P]>
        : T[P];
};

/*
Extract properties from an object of a certain type:

    type Person = {name: string, age: number, address: string},
    type StringFields = GetPropertiesByType<Person, string>
    // "name" | "address"

*/
export type GetPropertiesByType<T, FieldType> = {
    [Key in keyof T]: T[Key] extends FieldType ? Key : never;
}[keyof T];

/* Get inner type of array */
export type GetItemType<T> = T extends (infer U)[] ? U : never;
