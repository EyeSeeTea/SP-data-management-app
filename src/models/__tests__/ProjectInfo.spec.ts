import { getMockApi } from "../../types/d2-api";
import { getProject } from "./project-data";

const { api } = getMockApi();

test("getNodes", () => {
    const project = getProjectWithChanges();
    const nodes = project.info.getNodes();
    expect(nodes).toMatchSnapshot();
});

test("getAsString", () => {
    const project = getProjectWithChanges();
    const infoString = project.info.getAsString();
    expect(infoString).toMatchSnapshot();
});

test("getAsString on new project", () => {
    const project = getProject(api, {}).set("initialData", undefined);
    const infoString = project.info.getAsString();
    expect(infoString).toMatchSnapshot();
});

function getProjectWithChanges() {
    const project0 = getProject(api, {});
    const baseProject = project0.set("initialData", project0.data);
    const newSectors = baseProject.config.sectors.filter(s => s.code === "SECTOR_EDUCATION");
    const keepSector = baseProject.sectors[0];

    const project = baseProject
        .set("name", "Old name")
        .setSectors([keepSector].concat(newSectors))
        .setCovid19({
            dataElementsSet: baseProject.dataElementsSelection,
            sectorId: keepSector.id,
            dataElementIds: ["ik0ICagvIjm"],
            isSet: true,
        }).project;

    return project;
}
