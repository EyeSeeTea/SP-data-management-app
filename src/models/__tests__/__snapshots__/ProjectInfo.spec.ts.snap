// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAsString 1`] = `
"- Name: [UPDATED] MyProject -> Old name
- Description: -
- Award Number: 12345
- Subsequent Lettering: en
- Additional Designation (Funder, Location, Sector, etc): -
- Period dates: October 2018 - March 2019
- DART: No
- PARTNER: No
- Funders: funder-aOYJkeWdv2t, funder-yQKIZzBl22A
- Selected country: Bahamas
- Locations: loc-GsGG8967YDU
- Sharing:
  - Users: admin admin
  - User Groups: System Admin, Data Management Admin
- Sectors:
  - Agriculture:
    - de-B010200 - B010200
    - [UPDATED] de-P010100 - P010100 -> de-P010100 - P010100 [COVID-19]
    - [UPDATED] de-P010101 - P010101 -> de-P010101 - P010101 [COVID-19]
  - [ADDED] Education:
  - [REMOVED] Livelihoods:
    - de-B020200 - B020200 [MER]
    - de-P020300 - P020300
- Attached Files:"
`;

exports[`getAsString on new project 1`] = `
"- Name: MyProject
- Description: -
- Award Number: 12345
- Subsequent Lettering: en
- Additional Designation (Funder, Location, Sector, etc): -
- Period dates: October 2018 - March 2019
- DART: No
- PARTNER: No
- Funders: funder-aOYJkeWdv2t, funder-yQKIZzBl22A
- Selected country: Bahamas
- Locations: loc-GsGG8967YDU
- Sharing:
  - Users: admin admin
  - User Groups: System Admin, Data Management Admin
- Sectors:
  - Agriculture:
    - de-B010200 - B010200
    - de-P010100 - P010100
    - de-P010101 - P010101
  - Livelihoods:
    - de-B020200 - B020200 [MER]
    - de-P020300 - P020300
- Attached Files:"
`;

exports[`getNodes 1`] = `
Array [
  Object {
    "action": "updated",
    "name": "Name",
    "prevValue": "MyProject",
    "type": "field",
    "value": "Old name",
  },
  Object {
    "action": "none",
    "name": "Description",
    "prevValue": "",
    "type": "field",
    "value": "",
  },
  Object {
    "action": "none",
    "name": "Award Number",
    "prevValue": "12345",
    "type": "field",
    "value": "12345",
  },
  Object {
    "action": "none",
    "name": "Subsequent Lettering",
    "prevValue": "en",
    "type": "field",
    "value": "en",
  },
  Object {
    "action": "none",
    "name": "Additional Designation (Funder, Location, Sector, etc)",
    "prevValue": "",
    "type": "field",
    "value": "",
  },
  Object {
    "action": "none",
    "name": "Period dates",
    "prevValue": "October 2018 - March 2019",
    "type": "field",
    "value": "October 2018 - March 2019",
  },
  Object {
    "action": "none",
    "name": "DART",
    "prevValue": "No",
    "type": "field",
    "value": "No",
  },
  Object {
    "action": "none",
    "name": "PARTNER",
    "prevValue": "No",
    "type": "field",
    "value": "No",
  },
  Object {
    "action": "none",
    "name": "Funders",
    "prevValue": "funder-aOYJkeWdv2t, funder-yQKIZzBl22A",
    "type": "field",
    "value": "funder-aOYJkeWdv2t, funder-yQKIZzBl22A",
  },
  Object {
    "action": "none",
    "name": "Selected country",
    "prevValue": "Bahamas",
    "type": "field",
    "value": "Bahamas",
  },
  Object {
    "action": "none",
    "name": "Locations",
    "prevValue": "loc-GsGG8967YDU",
    "type": "field",
    "value": "loc-GsGG8967YDU",
  },
  Object {
    "action": "none",
    "children": Array [
      Object {
        "action": "none",
        "name": "Users",
        "prevValue": "admin admin",
        "type": "field",
        "value": "admin admin",
      },
      Object {
        "action": "none",
        "name": "User Groups",
        "prevValue": "System Admin, Data Management Admin",
        "type": "field",
        "value": "System Admin, Data Management Admin",
      },
    ],
    "title": "Sharing",
    "type": "section",
  },
  Object {
    "action": "none",
    "children": Array [
      Object {
        "action": "none",
        "children": Array [
          Object {
            "action": "none",
            "prevValue": "de-B010200 - B010200",
            "type": "value",
            "value": "de-B010200 - B010200",
          },
          Object {
            "action": "updated",
            "prevValue": "de-P010100 - P010100",
            "type": "value",
            "value": "de-P010100 - P010100 [COVID-19]",
          },
          Object {
            "action": "updated",
            "prevValue": "de-P010101 - P010101",
            "type": "value",
            "value": "de-P010101 - P010101 [COVID-19]",
          },
        ],
        "title": "Agriculture",
        "type": "section",
      },
      Object {
        "action": "added",
        "children": Array [],
        "title": "Education",
        "type": "section",
      },
      Object {
        "action": "removed",
        "children": Array [
          Object {
            "action": "none",
            "prevValue": "de-B020200 - B020200 [MER]",
            "type": "value",
            "value": "de-B020200 - B020200 [MER]",
          },
          Object {
            "action": "none",
            "prevValue": "de-P020300 - P020300",
            "type": "value",
            "value": "de-P020300 - P020300",
          },
        ],
        "title": "Livelihoods",
        "type": "section",
      },
    ],
    "title": "Sectors",
    "type": "section",
  },
  Object {
    "action": "none",
    "children": Array [],
    "title": "Attached Files",
    "type": "section",
  },
]
`;
